WT-7841 add "only unencrypted" checksum configuration, switch checksum default to "on"

keithbostic opened the pull request:
Add the "unencrypted" checksum configuration, to checksum only unencrypted blocks.

Switch default checksum configuration to "on".

@sueloverso, @agorrod hasn't approved the change to the default configuration, if/when that's done can you please do the review?

AlexeyAnisimovAU commented:
Since `uncompressed` option is not very safe, do we want to change the example code in [ex_all.c:622](https://github.com/wiredtiger/wiredtiger/blob/bb61aa3a8f9825639501e0723e65778d159c3b7d/examples/c/ex_all.c#L622) to `unencrypted`?

AlexeyAnisimovAU commented:
I'm not sure what the process of updating documentation is but [tune-checksum.dox](https://github.com/wiredtiger/wiredtiger/blob/develop/src/docs/tune-checksum.dox) requires an update too.

sueloverso commented on dist/api_data.py:
```diff
@@ -297,14 +297,21 @@ def __ge__(self, other):
+        detection is rarely the case for compression engines, specifically, none of the \c lz4,
+        \c snappy, \c zlib and \c zstd compression engines included in WiredTiger have checksum
+        support sufficient to allow turning off checksums.''',
+
```
Remove blank line

keithbostic replied:
Fixed in fddb5ac14.

sueloverso commented on dist/api_data.py:
```diff
@@ -297,14 +297,21 @@ def __ge__(self, other):
+        engines that detect block corruption. Corruption detection is often the case for
+        encryption engines, including the \c sodium engine included in WiredTiger. Corruption
+        detection is rarely the case for compression engines, specifically, none of the \c lz4,
+        \c snappy, \c zlib and \c zstd compression engines included in WiredTiger have checksum
```
Do we want to itemize these here? If/when the list changes someone will have to remember and find this place to change. I won't push hard as it is useful as written, only considering future maintenance.

keithbostic replied:
I think I'm missing your point, I don't see how itemization makes the problem better?

sueloverso replied:
I meant itemizing makes maintenance harder because if we stop including, say, snappy then someone has to remember to fix the checksum description which is not an obvious connection. Hopefully they'd grep.

keithbostic replied:
That paragraph only lives in the tuning doc, now, so I think it's likely to be found.

sueloverso commented on test/format/config.c:
```diff
@@ -499,15 +499,21 @@ config_checksum(void)
+        case 6: /* 10% */
             config_single("disk.checksum=uncompressed", false);
             break;
+        default: /* 40% */
```
10% for uncompressed and 40% for unencrypted seem skewed. Why not something more balanced?

keithbostic replied:
I think "on" and "unencrypted" are the most "likely correct" settings for applications, so I gave them each 40%. My opinion at the moment is that "uncompressed" is dangerous, I don't know of any compression engine we can trust.

sueloverso commented on test/format/config.c:
```diff
@@ -1368,9 +1374,11 @@ config_map_checksum(const char *s, u_int *vp)
         *vp = CHECKSUM_ON;
     else if (strcmp(s, "off") == 0)
-        *vp = CHECKSUM_ON;
+        *vp = CHECKSUM_OFF;
```
Yikes! I guess we weren't testing what we thought we were testing.

keithbostic replied:
Yeah, that was a shocker for me, too.

sueloverso reviewed (approved):
LGTM with a few minor comments.

keithbostic commented:
> Since `uncompressed` option is not very safe, do we want to change the example code in [ex_all.c:622](https://github.com/wiredtiger/wiredtiger/blob/bb61aa3a8f9825639501e0723e65778d159c3b7d/examples/c/ex_all.c#L622) to `unencrypted` or `on` to promote safer practice?

Nice catch, thanks!

Fixed in fddb5ac14: I rewrote the tuning page and changed the reference guide to point to the tuning page (instead of repeating the wording in both places). I left the example alone, I think it's OK as it is.

keithbostic commented:
> I'm not sure what the process of updating documentation is but [tune-checksum.dox](https://github.com/wiredtiger/wiredtiger/blob/develop/src/docs/tune-checksum.dox) and [upgrading.dox](https://github.com/wiredtiger/wiredtiger/blob/develop/src/docs/upgrading.dox) require an update too.

@AlexeyAnisimovAU, I plan to discard the "upgrading" section entirely before our next standalone release, it's really no longer applicable. Tagging @agorrod in case this is a surprise to him.

sueloverso reviewed (approved):
LGTM

AlexeyAnisimovAU reviewed (approved):
LGTM, nice comments and documentation, I enjoyed reading them!
